### ISSUE TITLE
[Proposal] Adopt the "purl" package URL spec to identify software packages

### ISSUE THREAD

**Issue by pombredanne** [Grafeas/Grafeas#65]
The initial [README](https://github.com/Grafeas/Grafeas/blob/b8ce70a4a88e18575e1d2d91912c156b078d5a8f/README.md#resource-urls) contains this section...

---------------

## Resource Urls
Component resource Urls need to be unique per resource as well as immutable. This will mean that the metadata associated with a resourceUrl will always be associated with exactly one component, and what is pointed at should never change. Content addressable resource urls are preferred. In the case with resources that cannot be immutable, a timestamp should be appended.

The following table provides examples one could use as resource urls for several component types:

Component Type|Identifier                                  |Example|
--------------|--------------------------------------------|-------|
|Debian       |deb://dist(optional):arch:name:version      |deb://lucid:i386:acl:2.2.49-2|
|Docker       |https://Namespace/name@sha256:<Checksum>    |https://gcr.io/scanning-customer/dockerimage@sha256:244fd47e07d1004f0aed9c156aa09083c82bf8944eceb67c946ff7430510a77b|
|Generic file |file://sha256:<Checksum>:name               |file://sha256:244fd47e07d1004f0aed9c156aa09083c82bf8944eceb67c946ff7430510a77b:foo.jar|
|Maven        |gav://group:artifact:version                |`gav://ant:ant:1.6.5`|
|NPM          |npm://package:version                       |npm://mocha:2.4.5|
|NuGet        |nuget://module:version                      |nuget://log4net:9.0.1|
|Python       |pip://package:version                       |pip://raven:5.13.0|
|RPM          |rpm://dist(optional):arch:name:version      |rpm://el6:i386:ImageMagick:6.7.2.7-4|

---------------

These examples provided are great but lack a proper definition and are not always consistent.
Several other tools and APIs use similar but different conventions to identify and locate packages. 

To better specify what these URLs are here and elsewhere (including in the https://github.com/nexB/scancode-toolkit and about.org projects and several other projects) I started to put down a draft spec at https://github.com/package-url/purl-spec/blob/042f1088bb17e0762e68b27daad30637ca44de4f/README.rst

It would be great for Grafeas to adopt such a spec too!

This is open for review at https://github.com/package-url/purl-spec/pull/1

PS: The original conversation was started here https://github.com/nexB/scancode-toolkit/issues/805

**Comment by pombredanne** [comment]
@R2wenD2 here we go as promised!

**Comment by R2wenD2** [comment]
Thanks @pombredanne I'll review and leave comments early next week. 
I don't think this changes the much for Grafeas - would just require the API implementation to handle and validate the revised formats.

**Comment by pombredanne** [comment]
Thanks!

> I don't think this changes the much for Grafeas - would just require the API implementation to handle and validate the revised formats.

Indeed, and that's part of the intent. 

FWIW there is a mention of a Go implementation in progress on the PR at https://github.com/package-url/purl-spec/pull/1#issuecomment-344409842

**Comment by brianf** [comment]
Maven has a few more coordinates that can come into play in some instances: packaging and classifier. The typical ordering looks like this:
groupId:artifactId:(optional)packaging:(optional)classifier:version

While GAV may be sufficient for many cases, there will be times where some attestation needs to apply to one classifier and not another, or one package and not another. We should expand the spec to allow the optional inclusion of these extra fields.

ref: https://maven.apache.org/pom.html#Maven_Coordinates

**Comment by pombredanne** [comment]
@brianf yes Maven has these and these are not the norm. They also in most derive from the same source, hence for these they would be treated as optional qualifiers in https://github.com/package-url/purl-spec/tree/a21a9ecbab171fe06d1ac88843bc3721a42dbed0 

    maven for Maven JARs and related artifacts
    - The default repository is maven.org
    - The group id is the namespace and the artifact id is the name
    - Known qualifiers keys are: classifier and packaging as defined in the POM documentation
    - Examples:
    `maven:org.apache.xmlgraphics/batik-anim@1.9.1`
    `maven:org.apache.xmlgraphics/batik-anim@1.9.1?packaging=sources`

**Comment by pombredanne** [comment]
@brianf It would be awesome if you want to chime in on the purl spec btw. The initial and a tad messy PR is there https://github.com/package-url/purl-spec/pull/1 and received comments from dpkg, nuget, spak, npm and several important package format maintainers and many others. Going forward a ticket is better.
Having a JVM implementation is also needed!

**Comment by brianf** [comment]
Is the intent of the purl spec to replace the one in the README? If not, how do they correlate?

**Comment by pombredanne** [comment]
@brianf that's the intent of this ticket to propose a replacement for the one of the README alright

**Comment by pombredanne** [comment]
The purl spec extends a bit beyond Grafeas as it also caters to other tools, DB and APIs such as libraries.io, victims, openshift analytics, aboutcode.org projects and a few more.

**Comment by lizrice** [comment]
The container image referencing in the purl spec looks good to me

**Comment by pombredanne** [comment]
@lizrice Thank you for the valuable feedback!

**Comment by R2wenD2** [comment]
I think everyone is happy with this proposal, we'll just need to add the verification to the reference implementation and documentation.

**Comment by vbatts** [comment]
woof this purl conversation has gone wild while I was out! So to be clear, https://github.com/package-url/purl-spec is a recent discussion and not a widely agreed-upon and implemented spec, right? 
I can follow the table, just want to make sure that it's a wider conversation.

**Comment by brianf** [comment]
If by recent / !wider you mean it's not widely implemented by tools, yes
that's true. The conversation has been going since mid November and this
seems to have good traction covering a large swath of packages out there.
Certainly from my POV, this is the best attempt at defining this I've ever
seen and think it fits perfectly into what Grafeas wants to be.

**Comment by R2wenD2** [comment]
@vbatts Do you have reservations about purl? If so I'd love to hear them. We want some kind of standard format for grafeas, so I basically +1 everything that @brianf said above.

**Comment by vbatts** [comment]
On 03/01/18 16:56 +0000, R2wenD2 wrote:
>@vbatts Do you have reservations about purl? If so I'd love to hear
>them. We want some kind of standard format for grafeas, so I basically
>+1 everything that @brianf said above.

faced with inventing something yourself, and working with others, it's
ideal to work together. 
Only reservation was just that it's not a widely embraced thing [yet].

**Comment by vbatts** [comment]
I don't intend to derail purl adoption. I think I've articulated my hesitation, which seems to be a general problem and not exclusive to purl. https://github.com/package-url/purl-spec/issues/29

**Comment by vbatts** [comment]
:heart:

**Comment by R2wenD2** [comment]
No problem. I think its still something we want to do. At this point, I expect that purl adoption means updating docs as well as the API implementation to parse for it.

**Comment by R2wenD2** [comment]
Next steps here:
1) Update docs
2) Incorporate the purl checks into our reference implementation!

**Comment by vbatts** [comment]
So I'm now learning the extent of the work that has gone into SWID ISO 19770-2, despite the pay-wall barrier. I'm very glad that purl-spec is more open, but we might should consider that many deployments may have requirements to use SWID over purl.
Purl-spec has a FAQ about this, https://github.com/package-url/purl-spec/wiki/FAQ#why-not-using-the-iso-19770-2-spec-for-swid-tags-instead-of-a-purl which I don't disagree with. Red Hat has participated in the SWID review, and is also a part urging/pressuring Tagvault to open the SWID spec up more.

**Comment by pombredanne** [comment]
@vbatts Just back from FOSDEM where I presented this https://fosdem.org/2018/schedule/event/purl/ (video and deck are there)
There was a clear preference from most everyone there to use a single scheme prefix eg. `pkg:npm:angular@1.2.3` such that this can be easily registered as an official RFC and many other good reasons so in this context we could have something like:
`pkg:swid:<whatever is a swid>` OR have a proper SWID URI or URL separately alright.
I wished SWID would not have been these walled garden centralized scheme, but that's not my battle to fight ;) Let's make a ticket on the purl side to have this included alright

**Comment by vbatts** [comment]
@pombredanne there is further conversation going regarding SWID. I am thinking that the prefixing may help. Further [this FAQ](https://github.com/package-url/purl-spec/wiki/FAQ#why-not-using-the-iso-19770-2-spec-for-swid-tags-instead-of-a-purl) really ought to be updated from "avoid this"-like language. As one [private] conversation included:
> I can understand the sentiment, but it is wrong in many respects. First, the
ISO specification is not for proprietary software. How can anyone come to
that conclusion. The C programming language is governed by ISO/IEC 9899:2011.
Does that mean that gcc is a proprietary compiler? I think not.

Having this distinguished prefixes, it'll let SWID do what it's intended for of specifying files (not just whole software packages, regardless of proprietary or FOSS), and purl can map collections/packages of software.

Make sense?
